### DISCUSSION TITLE
Q: Best Way for Device to Identify to Network?

### DISCUSSION THREAD

**Discussion by davidmcnabnz** [micropython/micropython#10845] [General]
Hi all,

Is there a recommended 'best practice' way for devices of identical hardware, and similar roles, to each identify themselves uniquely to a network?

I'm currently considering two main approaches:

1. Upload file-sets to each of the devices, that are identical except for a single device-specific `instance.py` config file. The device then uses values in instance.py to announce itself. This means the deployment process generates separate instance.py files, one per device, and uploads the 'shared' file set, plus each device's specific instance.py, versus
2. Use hardware-specifics such as WiFi chip MAC address (but I haven't yet figured out how to determine MAC address within the device's own MicroPython environment).

The purpose is that once the devices are identified to the network, they can receive instructions as to how each instance should configure its pins, react to input changes, report appropriately to the network. As such, the instances would be 'soft-configured' upon startup and registration.

Any ideas/suggestions/stories?

Cheers
David

**Comment by robert-hh** [comment]
Some ports provide `machine.unique_id`, which you could use to identify a device without loading personalized scripts to the board.

**Reply by davidmcnabnz** [reply]
Thanks, @robert-hh 

This simplifies the deployment upload, where the upload for each device can start (in Python) with:

```
DEVICE_TYPE = subprocess.getoutput("mpremote exec 'import uos; print(uos.uname().sysname)'")
DEVICE_ID = subprocess.getoutput("""mpremote exec 'import binascii,machine; print(binascii.hexlify(machine.unique_id()).decode("utf-8"))'""")
```
or (in Bash) with:
```
export DEVICE_TYPE=`mpremote exec "import uos; print(uos.uname().sysname)"`
export DEVICE_ID=`mpremote exec "import binascii,machine; print(binascii.hexlify(machine.unique_id()).decode('utf-8'))"`
````

And from that, the device id and type can be added to the server-side database, for soft-configuration upon startup and registration.

Much appreciated
David
